crates/optimism/payload/src/validator.rs

```diff
@@ -0,0 +1,85 @@
+//! Validates execution payload wrt Optimism consensus rules
+
+use alloc::sync::Arc;
+use alloy_rpc_types_engine::PayloadError;
+use derive_more::{Constructor, Deref};
+use op_alloy_rpc_types_engine::{OpExecutionData, OpPayloadError};
+use reth_optimism_forks::OpHardforks;
+use reth_payload_validator::{cancun, prague, shanghai};
+use reth_primitives::{Block, SealedBlock};
+use reth_primitives_traits::{Block as _, SignedTransaction};
+
+/// Execution payload validator.
+#[derive(Clone, Debug, Deref, Constructor)]
+pub struct OpExecutionPayloadValidator<ChainSpec> {
+    /// Chain spec to validate against.
+    #[deref]
+    inner: Arc<ChainSpec>,
+}
+
+impl<ChainSpec> OpExecutionPayloadValidator<ChainSpec>
+where
+    ChainSpec: OpHardforks,
+{
+    /// Returns reference to chain spec.
+    pub fn chain_spec(&self) -> &ChainSpec {
```

Comment:
I'd rather return the arc here so that we don't accidentally clone it
```suggestion
    pub fn chain_spec(&self) -> &Arc<ChainSpec> {
```